src/PrestaShopBundle/EventListener/ModuleActivatedListener.php

```diff
@@ -0,0 +1,198 @@
+            return;
+        }
+
+        $controller = $event->getController();
+
+        if (!is_array($controller)) {
+            return;
+        }
+
+        list($controllerObject, $methodName) = $controller;
+        $moduleActivated = $this->getAnnotation($controllerObject, $methodName);
+
+        if (null === $moduleActivated) {
+            return;
+        }
+
+        /** @var \Module $module */
+        $module = \Module::getInstanceByName($moduleActivated->getModuleName());
+        if (null === $module) {
+            return;
+        }
+
+        if (!$module->active) {
+            $this->showNotificationMessage($moduleActivated);
+            $url = $this->router->generate($moduleActivated->getRedirectRoute());
+
+            $event->setController(function () use ($url) {
+                return new RedirectResponse($url);
+            });
+        }
+    }
+
+    /**
+     * Send an error message when redirected, will only work on migrated pages.
+     *
+     * @param ModuleActivated $moduleActivated
+     */
+    private function showNotificationMessage(ModuleActivated $moduleActivated)
+    {
+        $this->session->getFlashBag()->add(
+            'error',
+            $this->translator->trans(
+                $moduleActivated->getMessage(),
+                [$moduleActivated->getModuleName()],
+                $moduleActivated->getDomain()
+            )
+        );
+    }
+
+    /**
+     * @param object $controllerObject
+     * @param string $methodName
+     *
+     * @return null|ModuleActivated
+     * @throws AnnotationException
+     * @throws \ReflectionException
+     */
+    private function getAnnotation($controllerObject, $methodName)
+    {
+        $tokenAnnotation = ModuleActivated::class;
+
+        $controllerClass = ClassUtils::getClass($controllerObject);
+        $classAnnotation = $this->annotationReader->getClassAnnotation(
+            new ReflectionClass($controllerClass),
+            $tokenAnnotation
+        );
+
+        if (null !== $classAnnotation && $classAnnotation instanceof ModuleActivated) {
+            $this->validateAnnotation($classAnnotation, $controllerClass);
+
+            return $classAnnotation;
+        }
+
+        $controllerReflectionObject = new ReflectionObject($controllerObject);
+        $reflectionMethod = $controllerReflectionObject->getMethod($methodName);
+
+        $annotation = $this->annotationReader->getMethodAnnotation($reflectionMethod, $tokenAnnotation);
+
+        if (null !== $annotation && $annotation instanceof ModuleActivated) {
+            $this->validateAnnotation($annotation, $controllerClass . '::' . $methodName);
+
+            return $annotation;
+        }
+
+        return null;
+    }
+
+    /**
+     * @param ModuleActivated $annotation
+     * @param string          $annotationPosition
+     * @throws AnnotationException
+     */
+    private function validateAnnotation(ModuleActivated $annotation, $annotationPosition)
```

Comment:
this event listener have more and more responsibilities: how about extracting the retrieval of the url into his own service.

So in the end you'll have something like:

```php
if (!is_array($controller)) {
      return;
}

$url = $this->urlFromAnnotationRetriever->retrieveUrl($controllerObject, $methodName);

if (!empty($url)) {
     $event->setController(function () use ($url) {
                return new RedirectResponse($url);
            });
}

Reply:
yes why not, but it will have to be a service specific to this Listener then?
what about the other listeners like DemoModeEnabledListener?